aslam_cv_pipeline/src/visual-npipeline.cc

```diff
@@ -122,66 +133,76 @@ std::shared_ptr<VisualNFrame> VisualNPipeline::getNext() {
 
 std::shared_ptr<VisualNFrame> VisualNPipeline::getNextImpl() {
   // Initialize the return value as null
-  std::shared_ptr<VisualNFrame> rval;
-  if(!completed_.empty()) {
-    // Get the oldest frame.
-    auto it = completed_.begin();
-    rval = it->second;
-    completed_.erase(it);
-    condition_not_full_.notify_all();
+  std::shared_ptr<VisualNFrame> nframe;
+  if (shutdown_ || completed_.empty()) {
+    return nframe;
   }
-  return rval;
+  // Get the oldest frame.
+  auto it = completed_.begin();
+  nframe = it->second;
+  completed_.erase(it);
+  condition_not_full_.notify_all();
+  return nframe;
+}
+
+void VisualNPipeline::processImageImpl(
+    size_t camera_index, const cv::Mat& image, int64_t timestamp) {
+  thread_pool_->enqueue(&VisualNPipeline::work, this, camera_index, image, timestamp);
 }
 
 std::shared_ptr<VisualNFrame> VisualNPipeline::getLatestAndClear() {
   std::shared_ptr<VisualNFrame> nframe;
   std::unique_lock<std::mutex> lock(mutex_);
-  if(!completed_.empty()) {
-    /// Get the latest frame.
-    auto it = completed_.rbegin();
-    nframe = it->second;
-    int64_t timestamp = it->first;
-    completed_.clear();
-    // Clear any processing frames older than this one.
-    auto pit = processing_.begin();
-    while(pit != processing_.end() && pit->first <= timestamp) {
-      pit = processing_.erase(pit);
-    }
+  if (shutdown_ || completed_.empty()) {
+    return nframe;
+  }
+  /// Get the latest frame.
+  auto it = completed_.rbegin();
```

Comment:
no auto.. and the variable names in this class are horrible :-) but just change them at your discretion